cloudfoundry/resource_cf_route_service_binding.go

```
@@ -63,48 +94,106 @@ func resourceRouteServiceBindingCreate(ctx context.Context, d *schema.ResourceDa
 	}
-	if !found {
-		d.SetId("")
-		return diag.Errorf("Route '%s' not found in service instance '%s'", routeID, serviceID)
-	}
 
-	d.Set("service_instance", serviceID)
-	d.Set("route", routeID)
+	d.Set("service_instance", routeServiceBinding.Relationships.ServiceInstance.Data.GUID)
+	d.Set("route", routeServiceBinding.Relationships.Route.Data.GUID)
+
 	return nil
 }
 
-func resourceRouteServiceBindingDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+func resourceRouteServiceBindingDelete(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
 	session := meta.(*managers.Session)
 
-	serviceID := d.Get("service_instance").(string)
-	routeID := d.Get("route").(string)
-	_, err := session.ClientV2.DeleteServiceBindingRoute(serviceID, routeID)
+	jobGUID, err := session.ClientGo.ServiceRouteBindings.Delete(context.Background(), d.Id())
+
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	if jobGUID != "" {
+		err = session.ClientGo.Jobs.PollComplete(context.Background(), jobGUID, nil)
+	}
 	return diag.FromErr(err)
 }
+
+func upgradeStateRouteServiceBindingStateV0toV1ChangeID(ctx context.Context, rawState map[string]any, meta any) (map[string]any, error) {
+	session := meta.(*managers.Session)
+
+	if len(rawState) == 0 {
+		log.Println("[DEBUG] Empty RouteServiceBinding; nothing to migrate.")
+		return rawState, nil
+	}
+
+	log.Printf("[DEBUG] Attributes before migration: %#v", rawState)
+	options := client.NewServiceRouteBindingListOptions()
+	options.ServiceInstanceGUIDs = client.Filter{Values: []string{rawState["service_instance"].(string)}}
+	options.RouteGUIDs = client.Filter{Values: []string{rawState["route"].(string)}}
+
+	routeBinding, err := session.ClientGo.ServiceRouteBindings.Single(context.Background(), options)
+
+	if err != nil {
+		log.Println("[DEBUG] Failed to migrate RouteServiceBinding id: did not find the route service binding.")
```

Maybe ignore error and set id to "" if the error is a 404 error

---

Agree and disagree? Maybe? It's debatable whether we should limit the state upgrader's responsibility to just upgrading the state and not modifying anything else.

---

I think that it's maybe best to set id to "" here, as it will behave as the Read function, so it's more consistent.

Let's take a situation where the resource was deleted outside of Terraform, so Terraform should recreate it on the next run. If we don't ignore the 404 error and set id to "" here, Terraform will fail at plan phase if state upgrade had to be done.
This looks like an unlikely scenario in my usage of the provider but I can imagine that some people might be interested in this

---

That's understandable. I'd prefer we double check the reason of the error to be "CF-ResourceNotFound" like in the Read case. I am afraid an unrelated network issue would cause the state to be dropped.

---

Good point, I'll add it

---

It's done & tested, could you check please ? @sleungcy